File: alloccount.py
#!/usr/bin/python
#
# strlen_count  Trace strlen() and print a frequency count of strings.
#               For Linux, uses BCC, eBPF. Embedded C.
#
# Written as a basic example of BCC and uprobes.
#
# Also see strlensnoop.
#
# Copyright 2016 Netflix, Inc.
# Licensed under the Apache License, Version 2.0 (the "License")

from __future__ import print_function
from bcc import BPF
from time import sleep
import os

# load BPF program
b = BPF(text="""
#include <uapi/linux/ptrace.h>

typedef struct blah {
    size_t ptr;
    pid_t pid;
} blah_t;

BPF_HASH(counts, blah_t);


int count(struct pt_regs *ctx) {
    pid_t pid = bpf_get_current_pid_tgid() & 0xffffffff;
    u64 zero = 0, *val;
    blah_t key = {};
    size_t ptr = PT_REGS_PARM1(ctx);
    key.ptr = ptr;
    key.pid = pid;
    val = counts.lookup_or_init(&key, &zero);
    (*val)++;
    return 0;
};
""")

b.attach_uprobe(name="/home/bork/.rbenv/versions/2.4.0/bin/ruby", sym="newobj_slowpath", fn_name="count")

# header
print("Tracing newobj_slowpath()... Hit Ctrl-C to end.")


counts = b.get_table("counts")

h = {}

import subprocess
from collections import namedtuple
Stats = namedtuple('Stats', ['count', 'pid', 'ptr'])

def update_cache(top):
    missing_ptrs = []
    missing_ptr_stats = []
    for stat in top:
        if stat.ptr == 0:
            continue
        if (stat.ptr, stat.pid) not in h:
            missing_ptrs.append(str(stat.ptr))
            missing_ptr_stats.append(stat)
    if len(missing_ptrs) == 0:
        return
    args = ["./target/debug/ruby-fork-test", str(missing_ptr_stats[0].pid)] + missing_ptrs
    # print(' '.join(args))
    out = subprocess.check_output(args)
    lines = out.split("\n")
    for l in lines:
        if len(l) == 0:
            continue
        (ptr, name, pid) = l.split()
        h[(int(ptr), int(pid))] = name



while True:
    os.system('clear')
    print("%30s | %s" % ("CLASS", "COUNT"))
    print("%30s | %s" % ("", ""))
    top = list(reversed(sorted([Stats(count=counts.get(key).value, ptr=key.ptr, pid=key.pid) for key in counts.keys()])))
    top = top[:20]
    if len(h) < 10:
        update_cache(top)
    for stat in top:
        s = h.get((stat.ptr, stat.pid), "?? %s" % stat.ptr)
        print("%30s | %s" % (s, stat.count))
    sleep(1)

File: src/bin/alloccount.rs
extern crate bcc_friendly;
extern crate byteorder;
extern crate libc;
extern crate failure;
use byteorder::{NativeEndian, ReadBytesExt};
use bcc_friendly::core::BPF;
use failure::Error;
use std::io::Cursor;

fn main() {
    do_main().unwrap();
}


fn do_main() -> Result<(), Error> {
    let code = "
#include <uapi/linux/ptrace.h>

typedef struct blah {
    size_t ptr;
    pid_t pid;
} blah_t;

BPF_HASH(counts, blah_t);

int count(struct pt_regs *ctx) {
    pid_t pid = bpf_get_current_pid_tgid() & 0xffffffff;
    u64 zero = 0, *val;
    blah_t key = {};
    size_t ptr = PT_REGS_PARM1(ctx);
    key.ptr = ptr;
    key.pid = pid;
    val = counts.lookup_or_init(&key, &zero);
    (*val)++;
    return 0;
};
    ";
    let mut module = BPF::new(code)?;
    let uprobe = module.load_uprobe("count".to_string())?;
    module.attach_uprobe("/home/bork/.rbenv/versions/2.4.0/bin/ruby".to_string(), "newobj_slowpath".to_string(), uprobe, -1)?;
    let mut table = module.table("counts");
    println!("{:?}", table.key_size());
    println!("{:?}", table.leaf_size());
    loop {
        std::thread::sleep(std::time::Duration::from_millis(1000));
        let iter = table.into_iter();
        for e in iter {
            let mut kcursor = Cursor::new(e.key);
            let ptr = kcursor.read_u64::<NativeEndian>().unwrap();
            let pid = kcursor.read_i32::<NativeEndian>().unwrap();
            let value = Cursor::new(e.value).read_u64::<NativeEndian>().unwrap();
            if value > 10 {
                println!("{:x} {:?} {:?}", ptr, pid, value);
            }
        }
    }
}

